apps/gdalalg_dispatcher.h

```diff
@@ -0,0 +1,242 @@
+/******************************************************************************
+ *
+ * Project:  GDAL
+ * Purpose:  gdal subcommand dispatcher
+ * Author:   Even Rouault <even dot rouault at spatialys.com>
+ *
+ ******************************************************************************
+ * Copyright (c) 2024, Even Rouault <even dot rouault at spatialys.com>
+ *
+ * SPDX-License-Identifier: MIT
+ ****************************************************************************/
+
+#ifndef GDALALG_DISPATCHER_INCLUDED
+#define GDALALG_DISPATCHER_INCLUDED
+
+#include "gdalalgorithm.h"
+
+#include "gdal_priv.h"
+
+//! @cond Doxygen_Suppress
+
+/************************************************************************/
+/*                       GDALDispatcherAlgorithm                        */
+/************************************************************************/
+
+template <class RasterDispatcher, class VectorDispatcher>
+class GDALDispatcherAlgorithm : public GDALAlgorithm
+{
+  public:
+    GDALDispatcherAlgorithm(const std::string &name,
+                            const std::string &description,
+                            const std::string &helpURL)
+        : GDALAlgorithm(name, description, helpURL),
+          m_rasterDispatcher(std::make_unique<RasterDispatcher>(
+              /* openForMixedRasterVector = */ true)),
+          m_vectorDispatcher(std::make_unique<VectorDispatcher>())
+    {
+        // A "info" dispacher command is a shortcut for something like
+        // "raster info", "vector info". Best to expose the latter.
+        SetDisplayInJSONUsage(false);
+    }
+
+    bool
+    ParseCommandLineArguments(const std::vector<std::string> &args) override;
+
+    std::string GetUsageForCLI(bool shortUsage,
+                               const UsageOptions &usageOptions) const override;
+
+  private:
+    std::unique_ptr<RasterDispatcher> m_rasterDispatcher{};
```

Comment:
I always do that, mostly to please -Weffc++ that would otherwise complain (one could argue that it shouldn't with C++ objects that have constructors). Better always initialize than miss one. I'd expect the compiler to optimize things